internal/otelarrow/admission2/boundedqueue.go

```diff
@@ -0,0 +1,177 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package admission2 // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/otelarrow/admission2"
+
+import (
+	"container/list"
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	grpccodes "google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var ErrTooMuchWaiting = status.Error(grpccodes.ResourceExhausted, "rejecting request, too much pending data")
+var ErrRequestTooLarge = status.Errorf(grpccodes.InvalidArgument, "rejecting request, request is too large")
+
+// BoundedQueue is a LIFO-oriented admission-controlled Queue.
+type BoundedQueue struct {
+	maxLimitAdmit uint64
+	maxLimitWait  uint64
+	tracer        trace.Tracer
+
+	// lock protects currentAdmitted, currentWaiting, and waiters
+
+	lock            sync.Mutex
+	currentAdmitted uint64
+	currentWaiting  uint64
+	waiters         *list.List // of *waiter
+}
+
+var _ Queue = &BoundedQueue{}
+
+// waiter is an item in the BoundedQueue waiters list.
+type waiter struct {
+	notify  N
+	pending uint64
+}
+
+// NewBoundedQueue returns a LIFO-oriented Queue implementation which
+// admits `maxLimitAdmit` bytes concurrently and allows up to
+// `maxLimitWait` bytes to wait for admission.
+func NewBoundedQueue(ts component.TelemetrySettings, maxLimitAdmit, maxLimitWait uint64) Queue {
+	return &BoundedQueue{
+		maxLimitAdmit: maxLimitAdmit,
+		maxLimitWait:  maxLimitWait,
+		waiters:       list.New(),
+		tracer:        ts.TracerProvider.Tracer("github.com/open-telemetry/opentelemetry-collector-contrib/internal/otelarrow"),
+	}
+}
+
+// acquireOrGetWaiter returns with three distinct conditions depending
+// on whether it was accepted, rejected, or asked to wait.
+//
+// - element=nil, error=nil: the fast success path
+// - element=nil, error=non-nil: the fast failure path
+// - element=non-nil, error=non-nil: the slow success path
+func (bq *BoundedQueue) acquireOrGetWaiter(pending uint64) (*list.Element, error) {
+	bq.lock.Lock()
+	defer bq.lock.Unlock()
+
+	if pending > bq.maxLimitAdmit {
+		// when the request will never succeed because it is
+		// individually over the total limit, fail fast.
+		return nil, ErrRequestTooLarge
+	}
+
+	if bq.currentAdmitted+pending <= bq.maxLimitAdmit {
+		// the fast success path.
+		bq.currentAdmitted += pending
+		return nil, nil
+	}
+
+	// since we were unable to admit, check if we can wait.
+	if bq.currentWaiting+pending > bq.maxLimitWait {
+		return nil, ErrTooMuchWaiting
+	}
```

Comment:
I suggest replacing currentAdmitted and currentWaiting with two atomic.Uint64 variables and keeping the lock only for the waiters list. This modification should be possible because these two counters are incremented and decremented independently. Benefits of such a modification:
- Faster fast-path for admission when capacity is available (no mutex acquisition needed).
- More granular locking, reducing contention.

For the fast-path, something like this could work:
```go
for {
    current := bq.currentAdmitted.Load()
    if current + pending <= bq.maxLimitAdmit {
        if bq.currentAdmitted.CompareAndSwap(current, current + pending) {
            return nil, nil
        }
        continue // retry if CAS failed
    }
    break 
}
```

And for the “can wait” scenario:

```go
waiting := bq.currentWaiting.Load()
if waiting + pending > bq.maxLimitWait {
    return nil, ErrTooMuchWaiting
}
```

Reply:
Introducing a `maxSpin` in the for-loop could also be useful to avoid an infinite loop under heavy load. If the number of spins becomes too high (e.g., more than 20-30), it likely indicates that the fast path is not practical.

Reply:
I have recorded this thread in https://github.com/open-telemetry/otel-arrow/issues/270. I would prefer to leave the code as-is, since it is derived from a well-tested code base that uses Mutexes.

I want to think about whether there is a race to be worried about between addWaiterLocked() and admitWaitersLocked() if we remove the lock from the fast path. 😅